File: c/2022/02/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif


int8_t report(char a, char b) {
    size_t aa = a-'A';
    size_t bb = b-'X';
    if (aa==bb) return 3 + bb + 1;
    else if ((aa+1)%3 == bb) return 6 + bb + 1;
    else return 0 + bb + 1;
}

int8_t result(char a, char b) {
    size_t aa = a-'A';
    size_t bb = b-'X';
    int8_t ret = 3 * bb;
    if (aa == 0) {
        ret += bb == 0 ? 3 : bb == 1 ? 1 : 2;
    } else if (aa == 1) {
        ret += bb == 0 ? 1 : bb == 1 ? 2 : 3;
    } else {
        ret += bb == 0 ? 2 : bb == 1 ? 3 : 1;
    }
    return ret;
}

void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int64_t i = 0; i < in->length; i++) {
        k += report(in->strings[i][0],in->strings[i][2]);
    }
    printf("Part 1: %ld\n", k);
    freeArrString(in);
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int64_t i = 0; i < in->length; i++) {
        k += result(in->strings[i][0],in->strings[i][2]);
    }
    printf("Part 2: %ld\n", k);
    freeArrString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2022/02/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_match(file, "([ABC]) ([XYZ])")

def pts(b):
    return 3 if b=='Z' else 2 if b=='Y' else 1

def report(a,b):
    if (a=='A' and b=='X') or (a=='B' and b=='Y') or (a=='C' and b=='Z'):
        return 3 + pts(b)
    elif (a=='A' and b=='Y') or (a=='B' and b=='Z') or (a=='C' and b=='X'):
        return 6 + pts(b)
    else:
        return 0 + pts(b)

def result(a,b):
    ret = 6 if b=='Z' else 3 if b=='Y' else 0
    if (a=='A'):
        ret += 3 if b=='X' else 1 if b=='Y' else 2
    elif (a=='B'):
        ret += 1 if b=='X' else 2 if b=='Y' else 3
    else:
        ret += 2 if b=='X' else 3 if b=='Y' else 1
    return ret

def part1():
    A = read_input()
    k = sum([report(a,b) for a,b in A])
    print("Part 1:",k)

def part2():
    A = read_input()
    k = sum([result(a,b) for a,b in A])
    print("Part 2:",k)

part1()
part2()